Updates for flavor reservation/virtualization

Mark-Powers opened the pull request:
- 6863a40..e894125 are cherry-picked from upstream, relating to flavor reservations
- 2a8f549 fixes issues with the flavor plugin and our fork
- b0fb41bf168d253eb0bf2968276e02375c99d6d0 adds lease update for extending a lease
- 3eb6c14 adds host randomization, and fixes host randomization within the baremetal host plugin too

priteau committed:
Fix InstanceReservations affinity nullable flag

Instance reservations can be created with the affinity parameter set to
True (affinity), False (anti-affinity) or null (no affinity requested).

Change Icf26d1b7dffe64f82d7084dcebb5df1c7c9f106d updated the database
schema with Alembic but missed updating the SQLAlchemy model.

Change-Id: Ic754692722a784b1cb47ffeebce58136a528ad14

JohnGarbutt committed:
Implement allocation_candidates for flavor-based reservation

This adds the flavor.instance plugin, where users request a reservation
similar to:

    openstack reservation lease create test --reservation \
        resource_type=flavor:instance,amount=2,\
        flavor_id='e26a4241-b83d-4516-8e0e-8ce2665d1966'

This uses lots of ideas from the existing instance reservation plugin,
and where possible leans on some of the existing logic, currently only
filter_hosts_by_reservation.

The plugin is not currently usable. There are follow-on patches that
implement reserve_resource and then finally update_reservation.

Change-Id: I7507e87908e16b313fd624810d6f43cb73dddafe

JohnGarbutt committed:
db: add compute resources and traits

For flavor-based reservation, we want to store all the resource class
inventory information from placement, along with all the trait
information for each host.

Change-Id: I154ee17190e1b2da77601699f03000cb670057af

JohnGarbutt committed:
Write inventory and traits into DB

Look up and store placement details when we add a compute host.

This assumes nothing changes. If it does you need to remove and add
hosts again to synchronise placement data.

Change-Id: I1fa9b40823bac94550a0fc1b83a9f6818936cea4

JohnGarbutt committed:
Add host_resource_inventory_get_all_per_host

This method is needed to query more details about each host when
building up possible flavor allocations.

Change-Id: Ifde07da9513089bde2e9353452e97c0648a756c6

JohnGarbutt committed:
Make flavor plugin usable

This implements everything except update_reservation, so you can create,
use and delete flavor-based reservations.

We are able to re-use the on_start and on_end implementation from the
instance plugin, along with its cleanup_resources implementation.

However, the flavors created for each reservation are different, because
we copy the details from the source flavor, then add the Blazar-specific
resource requests on top. This allows users to make use of pinned CPUs,
huge pages, and simple PCI device requests and vGPUs. We do not support
all flavor options, and these flavors will be rejected.

Change-Id: I5b2c9fded67ae7c593ea50ebef0f42b6c5856324

MoteHue committed:
:Only create a server group when affinity is set

Also adds docs so a user can find this behaviour easily.

Closes-Bug: #2071832
Change-Id: Iea3c4164f3e893c355ec21653f2e8f0e848941f0

MoteHue committed:
Document flavor-based instance reservations

Change-Id: Ia3470b5423a9757a8c43d2cbc2e90ffaca0e983b

priteau committed:
Delete server group only if one was created

Change-Id: Ib71c15ec989a86265b826ece00ad32f95e3292bb

priteau committed:
Enable flavor:instance plugin in DevStack

Change-Id: I98accb0f8caa652c6091f361c9eb7ff8279453da

Mark-Powers committed:
Fix flavor reservations for Chameleon Cloud

- Fix migration branches
- Add migration to allow affinity to be nullable
- Fix on_* events in flavor plugin
- Replace usage of context with lease (not sure context seems broken)
- Add before_end event to flavor reservations
- Fix db session code for Antelope
- Fix config assumptions

Mark-Powers committed:
Fix enforcement to pass flavor info in request

Mark-Powers committed:
Add description to flavor

Mark-Powers committed:
Fix flavor reservation before_end code

This is a customization in our fork, and so needed fixing.
Email relay is configured under the host plugin still based on our email script. See
https://github.com/ChameleonCloud/kolla-containers/commit/ee83055d89bf48a920723881c4ce2d007d41c1d8:

Mark-Powers committed:
Add option to disable physical:host reservation

Mark-Powers committed:
Fix unit tests

Mark-Powers committed:
Add update_reservation to flavor plugin

Mark-Powers committed:
Add randomization to flavor plugin

Mark-Powers committed:
Add "usage" information to host allocation tied to instance res

If a host allocation is tied to an instance reservation, we need to
include "usage" information to indicate how much of the host is
allocated. Otherwise, we have no idea what proportion of the host is
reserved.

msherman64 reviewed (commented):
I'm concerned about adding DB migrations without a major release.
It could lead to broken deployments if operators change container versions back and forth while still being in 2023.1.

Because of that, my preference may be to keep this branch (2023.1-kvm) separate until we bump to 2024.1, when operators will get an upgrade anyway, and shouldn't expect to roll back.

codyhammock reviewed (approved):
No notes, LGTM 👍🏻

Mark-Powers committed:
Add required flavor resource

Change-Id: I190dca0ba944d8131ad3f0aedbe5e43eac147192

Mark-Powers committed:
Revert "Delete instances running on reserved hosts when starting reservation"

This reverts commit 84fccdf89fe2d0c190f2f1e4fd622ba2cbd7a170.

Change-Id: Icb6dee82b955694e1e58e83e02cc0ebc2776acd0

Mark-Powers committed:
Update exception base class to use status 400

Change-Id: I5f681331088b3f0d61f2eddf6431a488c7cb850e

Mark-Powers committed:
Fix before_end actions with flavor reservations

Before before_end was set in allocation_candidates. This method
operates on a copy of the reservation, so the side effect of
setting before_end doesn't do anything. Instead we set it in
reserve_resource, which is actually creating the DB entry.

Change-Id: I73a48e7170cafaea4402c79a6ebf0627a7b55d8f

Mark-Powers committed:
Revert "Add required flavor resource"

This reverts commit ba1f398f81c4dc195fc1488e2cca2d964dd6a62e.

Change-Id: I99578e4c0a326a3dc409c3a99f5ae046938cdd8f

Mark-Powers committed:
Ignore flavor traits

Change-Id: I58652b6ba2e99bc06057ad0d78f8c1cd7705befa

Mark-Powers committed:
Add support for traits in flavor plugin

Previously, traits could not be included with a flavor reservation. This
commit implements checking flavor traits when querying available hosts.

This check is done via the placement API, in parallel where possible. As
a result, `_query_available_hosts` is slower when `resource_traits` is
non-empty.

Co-Authored-By: Michael Sherman <shermanm@uchicago.edu>
Change-Id: I5905d97d85955113d07b418cbe65fbafe132cd4b
Signed-off-by: Mark Powers <markpowers@uchicago.edu>

Mark-Powers committed:
Add support for flavor traits (#134)

* Add support for flavor traits

Change-Id: Id030e84e00ee734e0656f95c8d8a0c52d534f258

* Add key checking in trait check

Change-Id: Ic038fd61724f2f5cf6d0f749d1b7ed057eae8106

Mark-Powers committed:
Improve flavor trait checking

Previously checking traits was using get_trait_resource_providers()
makes n*m placement requests, where n is number of traits, and m
is number of hosts, sequentially. We parallelized n without
doing the same for m.

This new implementation makes m requests, in parallel.

Change-Id: I5905d97d85955113d07b418cbe65fbafe132cd4b

Mark-Powers committed:
Fix trait filtering

Change-Id: I68acbd60edb2207f1bda0b117452de6eaee1ac76

Mark-Powers committed:
Fix tests

Change-Id: Ic4768cc0ffbca2ddb61e78b9f064810131eed1c2

Mark-Powers committed:
Fix constraints

Change-Id: I5ac92ecfb912925094818ec90b99e39b2f582ab1

Mark-Powers committed:
Fix kvm unit tests

We added more DB calls for instance reservations in the host plugin,
and so we need the tests to account for this.

The logs throw a very confusing error when the DB is not properly
mocked in these methods about the DB file.

Change-Id: I8cf0e014b6ade6a9a2870eed85370d18f99a37f7

Mark-Powers committed:
Remove test --exclude from GHA

Change-Id: I9257b03d6f338457e15e6ffb4997043b8a709bd3

Mark-Powers committed:
Add host filter to flavor/host reservations

Change-Id: Icbebb86386d99a006a84b161ee05ec7c005acbae

Mark-Powers commented:
To support a hybrid site (VMs and BM), there is configuration for the following cases: 
- vm only
    - host.allow_reservation = False
- bm only
    - don't enable the flavor plugin
- vm and bm
    - `host.allow_reservation = True` -> permit host reservations
    - `host .filter_vm_hosts = True` -> don't permit vm host in a host res (require `hypervisor_type=ironic`)
    - `flavor.filter_ironic_hosts = True` -> don't permit bm host in a flavor res (forbid `hypervisor_type=ironic`)

Mark-Powers committed:
Update host monitor to use hypervisor instead of ID

In the past, blazar used to use the same host ID as the nova ID. We
stopped doing this, but never updated the nova host monitor code, since
we only use ironic hosts. The ironic host handler already uses this.

Change-Id: Ic435bc0538070bebe6d357fac443bffaab56bb3b

msherman64 commented on .github/workflows/test.yml:
```diff
@@ -33,6 +33,7 @@ jobs:
+            -c https://releases.openstack.org/constraints/upper/2023.1 \
             -r requirements.txt \
             -r test-requirements.txt \
             -c https://releases.openstack.org/constraints/upper/2023.1 \
```
super minor, but duplicate constraints file

Mark-Powers committed:
Fix gha

Change-Id: I15a24e4ff8231e5a141ea7ca3d04c752fc9e692e

Mark-Powers committed:
Fix flavor lease updates

This line got dropped somehow in the git history during a rebase.
Added some new tests as well.

Change-Id: Id398707ec20c9e14e2d7bd5c1f11ad3008ad32a3

Mark-Powers committed:
Fix flavor plugin _max_usages

Fixes the max usage calculation on a host to use the actual number
of allocations on that host. Previously, this used 'amount' as the
number of allocations, which is not correct in the case where a
reservation has allocations spread across multiple hosts.

Change-Id: I30d39629050accb81392b3555a338caed0d7d8aa

Mark-Powers committed:
Ignore flavor reservations when updating hosts

Users don't specify resource/hypervisor properties when making flavor
reservations, so there is no need to check them.

Change-Id: I4f122093cd4de4e358cbe8a5cfe7dbfb8ca1354c

Mark-Powers committed:
Fix lease update project ID check

Change-Id: Iee4884851fa5867aefdca92f7182b86df150a38b

Mark-Powers committed:
Add cleaning time to flavor branch

Change-Id: I645c632b317893af2e8dfa7324b0d89ef5e6db91

Mark-Powers committed:
Sum resource inventory for a host across child RPs

The previous implementation only looked at the top level RP when storing
the resource inventory. This is an issue since VGPUs are set on child
resource providers.

This change sums the total resources for each host across all RPs in the
tree, starting with the host's resource provider.

Change-Id: Ib3cd3b5d008a60107cad2e9ab6b7e918684195a3

Mark-Powers committed:
Add warning for different max_unit

Change-Id: Ibad26f12651b49ecc2aa65ca2feb20d23095209a

Mark-Powers committed:
Update lease usage

Change-Id: I358c51a9a9d6dc5321b5bf79350da037496893a1

Mark-Powers committed:
Fix extras handling in usage

Change-Id: Ic9b974b7978c0b2967692f991a15534182490b12

Mark-Powers committed:
Permit admins to make host reservations

Mark-Powers committed:
Add lease ID to flavor description

Mark-Powers committed:
Add flavor availability API

Mark-Powers committed:
Remove unused code
